File: 01/day_1.kt
fun main(args: Array<String>) {
	println("Part 1: ${partOneSolve(input+input[0], 0)}")
    
    val shift = (input + input).slice(IntRange(input.length/2, input.length+input.length/2-1))
    println("Part 2: ${partTwoSolve(input, shift, 0)}")
}

fun partOneSolve(input: String, acc: Int): Int {
    if (input.length < 2) {
       return acc
    }
    
    if (input[0].compareTo(input[1]) == 0) {
    	return partOneSolve(input.drop(1), acc + (input[0].toInt() - 48))
    }
    
    return partOneSolve(input.drop(1), acc)
}

fun partTwoSolve(input: String, shift: String, acc: Int): Int {
    if (input.length == 0) {
        return acc
    }
    
    if (input[0].compareTo(shift[0]) == 0) {
        return partTwoSolve(input.drop(1), shift.drop(1), acc + (input[0].toInt() - 48))
    }
    
    return partTwoSolve(input.drop(1), shift.drop(1), acc)
}

File: 01/day_1.go
package main

import (
	"fmt"
	"io/ioutil"
	"strconv"
	"strings"
)

func main() {
	bytes, err := ioutil.ReadFile("input.txt")
	if err != nil {
		panic(err)
	}

	input := strings.Split(strings.TrimSpace(string(bytes)), "")

	fmt.Printf("Part 1: %d\n", partOneSolve(append(input, input[0]), 0))

	length := len(input)
	shift := append(input, input...)[length/2 : length/2+length]
	fmt.Printf("Part 2: %d\n", partTwoSolve(input, shift, 0))
}

func partOneSolve(input []string, acc int) int {
	if len(input) < 2 {
		return acc
	}

	if input[0] == input[1] {
		val, err := strconv.Atoi(input[0])
		if err != nil {
			panic(err)
		}

		acc += val
	}

	return partOneSolve(input[1:], acc)
}

func partTwoSolve(input []string, shift []string, acc int) int {
	if len(input) == 0 {
		return acc
	}

	if input[0] == shift[0] {
		val, err := strconv.Atoi(input[0])
		if err != nil {
			panic(err)
		}

		acc += val
	}

	return partTwoSolve(input[1:], shift[1:], acc)
}
